dask/dataframe/io/hdf.py

```diff
@@ -506,24 +410,141 @@ def read_hdf(
             "When assuming pre-partitioned data, data must be "
             "read in its entirety using the same chunksizes"
         )
-    from ..multi import concat
-
-    return concat(
-        [
-            _read_single_hdf(
-                path,
-                key,
-                start=start,
-                stop=stop,
-                columns=columns,
-                chunksize=chunksize,
-                sorted_index=sorted_index,
-                lock=lock,
-                mode=mode,
-            )
-            for path in paths
-        ]
+
+    # Build metadata
+    with pd.HDFStore(paths[0], mode=mode) as hdf:
+        meta_key = _expand_key(key, hdf)[0]
+    meta = pd.read_hdf(paths[0], meta_key, mode=mode, stop=0)
+    if columns is not None:
+        meta = meta[columns]
+
+    # Common kwargs
+    if meta.ndim == 1:
+        common_kwargs = {"name": meta.name, "mode": mode}
+    else:
+        common_kwargs = {"mode": mode}
+
+    # Build parts
+    parts, divisions = build_parts(
+        paths, key, start, stop, chunksize, sorted_index, mode
+    )
+
+    # Construct Layer and Collection
+    label = "read-hdf-"
+    name = label + tokenize(paths, key, start, stop, sorted_index, chunksize, mode)
+    if len(divisions) != len(parts) + 1:
+        divisions = [None] * (len(parts) + 1)
+    layer = DataFrameIOLayer(
+        name,
+        columns,
+        parts,
+        HDFFunctionWrapper(columns, meta.ndim, lock, common_kwargs),
+        label=label,
     )
+    graph = HighLevelGraph({name: layer}, {name: set()})
+    return new_dd_object(graph, name, meta, divisions)
+
+
+def build_parts(paths, key, start, stop, chunksize, sorted_index, mode):
+    """
+    Build the list of partition inputs and divisions for read_hdf
+    """
+    parts = []
+    global_divisions = []
+    for path in paths:
+
+        keys, stops, divisions = get_keys_stops_divisions(
+            path, key, stop, sorted_index, chunksize, mode
+        )
+
+        for k, stop, division in zip(keys, stops, divisions):
+
+            if division and global_divisions:
+                global_divisions = global_divisions[:-1] + division
+            elif division:
+                global_divisions = division
```

Comment:
Ouch, this is tricky. Can we make the divisions global to begin with?

Reply:
Since this PR is just a refactor, most of the actual logic is just a copy of what already existed (iow, this is not my code, and I haven't thought through it carefully).  With that said, I can try to revist this later and simplify the logic a bit.

Reply:
Yeah, I can appreciate not wanting to mess with that logic too much